### PR TITLE
rust: module: improve parsing and errors

### PR DESCRIPTION
**Pull request by ojeda**
- Simplified code.

  - Repeated keys are not allowed anymore, e.g.

        module! {
            type: RustMinimal,
            type: RustMinimal,
        ...

  - Better error messages:

        error: proc macro panicked
          --> ../samples/rust/rust_minimal.rs:10:1
           |
        10 | / module! {
        11 | |     type: RustMinimal,
        12 | |     name: b"rust_minimal",
        13 | |     description: b"Rust minimal sample",
        14 | |     author: b"Rust for Linux Contributors",
        15 | |     license: b"GPL v2",
        16 | | }
           | |_^
           |
           = help: message: Keys are not ordered as expected. Order them like: ["type", "name", "author", "description", "license"]

        error: proc macro panicked
          --> ../samples/rust/rust_minimal.rs:10:1
           |
        10 | / module! {
        11 | |     typo: RustMinimal,
        12 | |     name: b"rust_minimal",
        13 | |     author: b"Rust for Linux Contributors",
        14 | |     description: b"Rust minimal sample",
        15 | |     license: b"GPL v2",
        16 | | }
           | |_^
           |
           = help: message: Unknown key "typo". Valid keys are: ["type", "name", "author", "description", "license", "alias", "alias_rtnl_link", "params"].

        error: proc macro panicked
          --> ../samples/rust/rust_minimal.rs:10:1
           |
        10 | / module! {
        11 | |     type: RustMinimal,
        12 | |     name: b"rust_minimal",
        13 | |     author: b"Rust for Linux Contributors",
        14 | |     description: b"Rust minimal sample",
        15 | | }
           | |_^
           |
           = help: message: Missing required key "license".

        error: proc macro panicked
          --> ../samples/rust/rust_minimal.rs:10:1
           |
        10 | / module! {
        11 | |     type: RustMinimal,
        12 | |     name: b"rust_minimal",
        13 | |     license: b"GPL v2",
        14 | |     license: b"GPL v2",
        15 | | }
           | |_^
           |
           = help: message: Duplicated key "license". Keys can only be specified once.

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

### CODE REVIEW

**Comment by ojeda** [comment]
Related: #215 and #234.

**Comment by bjorn3** [comment]
> ```
> error: proc macro panicked
>   --> ../samples/rust/rust_minimal.rs:10:1
>    |
> 10 | / module! {
> 11 | |     typo: RustMinimal,
> 12 | |     name: b"rust_minimal",
> 13 | |     author: b"Rust for Linux Contributors",
> 14 | |     description: b"Rust minimal sample",
> 15 | |     license: b"GPL v2",
> 16 | | }
>    | |_^
>    |
>    = help: message: Unknown key "typo". Valid keys are: ["type", "name", "author", "description", "license", "alias", "alias_rtnl_link", "params"].
> ```

I think you can improve this error message by expanding to a `compile_error!()` macro invocation with the span set to that of `typo`.

File: rust/module.rs [hunk 1]
```diff
@@ -321,81 +321,75 @@ struct ModuleInfo {
...
-        // The first allowed position is 1, to prevent setting something twice.
-        let mut pos = 0u8;
+
+        const EXPECTED_KEYS: &[&str] = &[
+            "type",
+            "name",
+            "author",
+            "description",
+            "license",
+            "alias",
```
Comment (kloenk) [inline rust/module.rs:331]: I do not yet fully understand if alias can be there multiple times. We should evaluate that somedays.

**Review by kloenk — APPROVED** [review]
LGTM.

I'm not sure if `compile_error!` would make it better. Probably it would make the output more readable for the specific error message, but It also would mean that the focus of the error get's shifted away from the proc_macro. So I'm fine with either ways.

File: rust/module.rs (hunk 1 above)
Comment (ojeda) [reply rust/module.rs:331]: Yeah, there are a few things regarding this -- let me create an issue.

**Comment by ojeda** [comment]
> I think you can improve this error message by expanding to a `compile_error!()` macro invocation with the span set to that of `typo`.

Tried that hack... it looks quite nice:

```
error: Duplicated key "type". Keys can only be specified once.
  --> ../samples/rust/rust_minimal.rs:12:5
   |
12 |     type: RustMinimal,
   |     ^^^^
```

Having said that, Rust should *really* stabilize something for proper error reporting in proc macros. Not sure if it is worth the pain at the moment...

**Comment by ojeda** [comment]
> I'm not sure if compile_error! would make it better. Probably it would make the output more readable for the specific error message, but It also would mean that the focus of the error get's shifted away from the proc_macro. So I'm fine with either ways.

Yeah, good point, I would need to mention `module!` in the message.

**Pull request merged**